language/src/regent/optimize_index_launches.t

```diff
@@ -1154,18 +1163,127 @@ local function optimize_loop_body(cx, node, report_pass, report_fail)
     end
   end
 
-  report_pass("loop optimization succeeded")
+  if needs_dynamic_check then
+    report_pass("static loop optimization failed, emitting dynamic check")
+  else
+    report_pass("loop optimization succeeded")
+  end
+
   return {
     preamble = preamble,
     call = call,
     reduce_lhs = reduce_lhs,
     reduce_op = reduce_op,
     args_provably = args_provably,
     free_variables = free_vars,
-    loop_variables = loop_vars
+    loop_variables = loop_vars,
+    needs_dynamic_check = needs_dynamic_check
   }
 end
 
+function init_bitmask_false(bitmask, volume)
+  local stats = terralib.newlist()
+
+  local idx = std.newsymbol(int32, "i")
+  local bitmask_i = util.mk_expr_index_access(util.mk_expr_id_rawref(bitmask), util.mk_expr_id(idx), std.rawref(&bool))
+  local assign = util.mk_stat_assignment(bitmask_i, util.mk_expr_constant(false, bool))
+  stats:insert(assign)
+
+  local values = terralib.newlist()
+  values:insert(util.mk_expr_constant(0, int32))
+  values:insert(util.mk_expr_id(volume))
+
+  return util.mk_stat_for_num(idx, values, util.mk_block(stats))
+end
+
+function get_check_stats(preamble, value, index_expr, bitmask, conflict, pos, loop_var, volume)
+  local stats = terralib.newlist()
+
+  -- Compute value = index_expr(i)
+  preamble:map(function(stat) stats:insert(stat) end)
+  stats:insert(util.mk_stat_assignment(util.mk_expr_id_rawref(value), index_expr))
+
+  local then_block = terralib.newlist()
+
+  local bitmask_value = util.mk_expr_index_access(util.mk_expr_id(bitmask), util.mk_expr_id(value), std.rawref(&bool))
+  local conflict_assign = util.mk_stat_assignment(util.mk_expr_id_rawref(conflict), bitmask_value)
```

Comment:
This is a minor preference, but since the variable is used exactly once on the following line, I'd be inclined not to create a variable and to instead just pass it straight to the `:insert()` call.

Reply:
Fixed.